moveit_ros/trajectory_cache/include/moveit/trajectory_cache/trajectory_cache.hpp

```diff
@@ -212,72 +288,76 @@ class TrajectoryCache
 
   /**
-   * @brief Inserts a trajectory into the database if it is the best matching trajectory seen so far.
-   *
-   * Trajectories are matched based off their start and goal states.
-   * And are considered "better" if they are higher priority in the sorting order specified by `sort_by` than another
-   * exactly matching trajectory.
-   *
-   * A trajectory is "exactly matching" if its start and goal are close enough to another trajectory.
-   * The tolerance for this depends on the `exact_match_precision` arg passed in init().
-   * @see init()
+   * @brief Inserts a trajectory into the database, with user-specified insert policy.
    *
    * Optionally deletes all worse trajectories by default to prune the cache.
    *
+   * @see FeaturesInterface<FeatureSourceT>
+   * @see CacheInsertPolicyInterface<KeyT, ValueT, CacheEntryT>
+   *
    * @param[in] move_group. The manipulator move group, used to get its state.
    * @param[in] cache_namespace. A namespace to separate cache entries by. The name of the robot is a good choice.
-   * @param[in] plan_request. The motion plan request to key the cache with.
-   * @param[in] trajectory. The trajectory to put.
-   * @param[in] execution_time_s. The execution time of the trajectory, in seconds.
-   * @param[in] planning_time_s. How long the trajectory took to plan, in seconds.
-   * @param[in] prune_worse_trajectories. If true, will prune the cache by deleting all cache entries that match the
-   * `plan_request` exactly, if they are worse than the `trajectory`, even if it was not put.
-   * @returns true if the trajectory was the best seen yet and hence put into the cache.
+   * @param[in] plan_request. The motion plan request to extract features from to key the cache with.
+   * @param[in] plan. The plan containing the trajectory to insert.
+   * @param[in,out] cache_insert_policy. The cache insert policy to use. Will determine what features can be used to
+   * fetch entries, and pruning and insertion logic. Will be reset at the end of the call.
+   * @param[in] prune_worse_trajectories. If true, will prune the cache according to the `cache_insert_policy`'s pruning
+   * logic.
+   * @param[in] additional_features. Additional features to key the cache with. Must not intersect with the set of
+   * features supported by the `cache_insert_policy`.
+   * @returns True if the trajectory was inserted into the cache.
    */
   bool insertTrajectory(const moveit::planning_interface::MoveGroupInterface& move_group,
                         const std::string& cache_namespace, const moveit_msgs::msg::MotionPlanRequest& plan_request,
-                        const moveit_msgs::msg::RobotTrajectory& trajectory, double execution_time_s,
-                        double planning_time_s, bool prune_worse_trajectories = true);
+                        const moveit::planning_interface::MoveGroupInterface::Plan& plan,
+                        CacheInsertPolicyInterface<moveit_msgs::msg::MotionPlanRequest,
```

Comment:
The CacheInsertPolicyInterface is passed by reference so it can be reset() at the end of the call, which means we need a non-const reference.

This is relevant since a cache insert policy might want to hold state during the insert call to help in pruning or the insert decision.

Case in point: The best seen execution time policy keeps track of best seen execution time, but has its internal tracking variable reset at the end of the call. [Code](https://github.com/moveit/moveit2/blob/a8ec0ac75e4eadf11b027ec84f98a47e4d503d49/moveit_ros/trajectory_cache/src/cache_insert_policies/best_seen_execution_time_policy.cpp)